examples/timebase_master.py

```diff
@@ -0,0 +1,148 @@
+            pos.valid |= jack._lib.JackPositionBBT
+
+            minutes = pos.frame / (pos.frame_rate * 60.0)
+            abs_tick = minutes * pos.beats_per_minute * pos.ticks_per_beat
+            abs_beat = abs_tick / self.ticks_per_beat
+
+            pos.bar = int(abs_beat / pos.beats_per_bar)
+            pos.beat = int(abs_beat - (pos.bar * pos.beats_per_bar) + 1)
+            pos.tick = int(abs_tick - (abs_beat * pos.ticks_per_beat))
+            pos.bar_start_tick = pos.bar * pos.beats_per_bar * pos.ticks_per_beat
+            pos.bar += 1  # adjust start to bar 1
+        else:
+            # Compute BBT info based on previous period.
+            pos.tick += int(nframes * pos.ticks_per_beat *
+                            pos.beats_per_minute / (pos.frame_rate * 60))
+
+            while pos.tick >= pos.ticks_per_beat:
+                pos.tick -= int(pos.ticks_per_beat)
+                pos.beat += 1
+
+                if pos.beat > pos.beats_per_bar:
+                    pos.beat = 1
+                    pos.bar += 1
+                    pos.bar_start_tick += pos.beats_per_bar * pos.ticks_per_beat
+
+                    if self.debug:
+                        print("Pos:", jack.position2dict(pos))
+
+    def become_timebase_master(self, conditional=None):
+        return self.set_timebase_callback(self._tb_callback, conditional
+                                          if conditional is not None
+                                          else self.conditional)
+
+
+def main(args=None):
+    ap = argparse.ArgumentParser(description=__doc__)
+    ap.add_argument(
+        '-d', '--debug',
+        action='store_true',
+        help="Enable debug messages")
+    ap.add_argument(
+        '-c', '--conditional',
+        action='store_true',
+        help="Exit if another timebase master is already active")
+    ap.add_argument(
+        '-n', '--client-name',
+        metavar='NAME',
+        default='timebase',
+        help="JACK client name (default: %(default)s)")
+    ap.add_argument(
+        '-m', '--meter',
+        default='4/4',
+        help="Meter as <beats-per-bar>/<beat-type> (default: %(default)s)")
+    ap.add_argument(
+        '-t', '--ticks-per-beat',
+        type=int,
+        metavar='NUM',
+        default=1920,
+        help="Ticks per beat (default: %(default)s)")
+    ap.add_argument(
+        'tempo',
+        nargs='?',
+        type=float,
+        default=120.0,
+        help="Tempo in beats per minute (default: %(default)s)")
+
+    args = ap.parse_args(args if args is not None else sys.argv[1:])
+
+    try:
+        beats_per_bar, beat_type = (int(x) for x in args.meter.split('/', 1))
+    except (TypeError, ValueError):
+        print("Error: invalid meter: %s\n" % args.meter)
+        ap.print_help()
+        return 2
+
+    try:
+        tbmaster = TimebaseMasterClient(
+            args.client_name,
+            bpm=max(0.0, min(300.0, args.tempo)),
```

Comment:
Why are you silently limiting this instead of raising an error?

Why are you limiting it at all?

I didn't know that, but JACK seems to allow negative BPM values ...

Reply:
Because BPM values above 300 aren't very musically useful, IMHO.

I could have implemented a custom `argparse` argument checker, but it seems overkill here. It's just an example, after all.

And how would a negative BPM work? Come to think of it, 0 BPM makes no sense either.

I changed the clamping to the range `0.1 - 300.0` and updated the param help to show the min/max value in da755992a6c251d2503c6e806de7d91f56d5efca.

Reply:
> aren't very musically useful

maybe, maybe not.

I think it is much more interesting to see how JACK reacts to certain values, than what restrictions were added in this example.

> I could have implemented a custom `argparse` argument checker

I wouldn't check the range at all.

> And how would a negative BPM work? Come to think of it, 0 BPM makes no sense either.

I don't know any of this.

But I would certainly be interested how JACK handles those values, and if you are limiting them in the example, I can't easily find out.